icicle/appUtils/poseidon/poseidon.cu

```diff
@@ -0,0 +1,256 @@
+#include "poseidon.cuh"
+
+// Used in matrix multiplication
+extern __shared__ scalar_t shared_states[];
+
+__global__ void prepare_poseidon_states(scalar_t * inp, scalar_t * states, size_t number_of_states, scalar_t domain_tag, const PoseidonConfiguration config) {
+    int idx = (blockIdx.x * blockDim.x) + threadIdx.x;
+    int state_number = idx / config.t;
+    if (state_number >= number_of_states) {
+        return;
+    }
+    int element_number = idx % config.t;
+
+    scalar_t prepared_element;
+
+    // Domain separation
+    if (element_number == 0) {
+        prepared_element = domain_tag;
+    } else {
+        prepared_element = inp[state_number * (config.t - 1) + element_number - 1];
+    }
+
+    // Add pre-round constant
+    prepared_element = prepared_element + config.round_constants[element_number];
+
+    // Store element in state
+    states[idx] = prepared_element;
+}
+
+__device__ __forceinline__ scalar_t sbox_alpha_five(scalar_t element) {
+    scalar_t result = element * element;
+    result = result * result;
+    return result * element;
+}
+
+__device__ scalar_t vecs_mul_matrix(scalar_t element, scalar_t * matrix, int element_number, int vec_number, int size) {
+    shared_states[threadIdx.x] = element;
+    __syncthreads();
+
+    element = scalar_t::zero();
+    for (int i = 0; i < size; i++) {
+        element = element + (shared_states[vec_number * size + i] * matrix[i * size + element_number]);
```

Comment:
Ideally, the multiplications here shouldn't be followed by reductions, with only one reduction performed in the end after the loop. We have `scalar_t::mul_wide` for the multiplication without reduction, but lack methods to accumulate and reduce `element` in the end. I'll add this functionality in the next PR and we'll be able to make this significantly faster.

Reply:
Fixed in 24222badb6d0db93bc0f0645ef7103beb4a6fa30